src/mpl2/src/clusterEngine.cpp

```diff
@@ -0,0 +1,2160 @@
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// * Neither the name of the copyright holder nor the names of its
+//   contributors may be used to endorse or promote products derived from
+//   this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+///////////////////////////////////////////////////////////////////////////////
+
+#include "clusterEngine.h"
+
+#include <queue>
+
+#include "db_sta/dbNetwork.hh"
+#include "object.h"
+#include "par/PartitionMgr.h"
+#include "sta/Liberty.hh"
+
+namespace mpl2 {
+using utl::MPL;
+
+ClusteringEngine::ClusteringEngine(odb::dbBlock* block,
+                                   sta::dbNetwork* network,
+                                   utl::Logger* logger,
+                                   par::PartitionMgr* triton_part)
+    : block_(block),
+      network_(network),
+      logger_(logger),
+      triton_part_(triton_part),
+      level_(0),
+      id_(0)
+{
+}
+
+void ClusteringEngine::run()
+{
+  initTree();
+  setBaseThresholds();
+
+  createIOClusters();
+  createDataFlow();
+
+  if (design_metrics_->getNumStdCell() == 0) {
+    logger_->warn(MPL, 25, "Design has no standard cells!");
+    tree_->has_std_cells = false;
+    treatEachMacroAsSingleCluster();
+  } else {
+    multilevelAutocluster(tree_->root);
+
+    std::vector<std::vector<Cluster*>> mixed_leaves;
+    fetchMixedLeaves(tree_->root, mixed_leaves);
+    breakMixedLeaves(mixed_leaves);
+  }
+
+  if (logger_->debugCheck(MPL, "multilevel_autoclustering", 1)) {
+    logger_->report("\nPrint Physical Hierarchy\n");
+    printPhysicalHierarchyTree(tree_->root, 0);
+  }
+
+  // Map the macros in each cluster to their HardMacro objects
+  for (auto& [cluster_id, cluster] : tree_->maps.id_to_cluster) {
+    mapMacroInCluster2HardMacro(cluster);
+  }
+}
+
+void ClusteringEngine::setDesignMetrics(Metrics* design_metrics)
+{
+  design_metrics_ = design_metrics;
+}
+
+void ClusteringEngine::setTree(PhysicalHierarchy* tree)
+{
+  tree_ = tree;
+}
+
+// Fetch the design's logical data
+void ClusteringEngine::fetchDesignMetrics()
```

Comment:
This is an odd name as nothing is fetch and it is confusing with computeMetrics.  What is the conceptual operation here?

Reply:
The conceptual operation would be to populate the Metrics object stored in HierRTLMP with the design's data.

Reply:
computeDesignMetrics perhaps?  Though computeMetrics might need a more specific name to distinguish what it does from this.

Reply:
Within this logic, computeMetrics should be computeModuleMetrics. I usually like to use computeSomething for getters that return an object, that's why I thought of fetch. But I think computeDesignMetrics and computeModuleMetrics seems good. What do you think?

Reply:
fetch sounds like a getter to me though I always use get naming for that purpose.  I can see expecting compute to return a result.  

I think the problem is that this method is a mixture of things: calling computeMetrics, reporting, assigning macro_with_halo_area , and error checking.   Making it more single purpose would probably ease naming.

Reply:
After thinking about this, I thought it would be better to remove this fetchDesignMetrics as we'll actually populate the Metrics when doing
```
design_metrics_ = computeModuleMetrics(block_->getTopModule());
```
 So we'll have an init() function to check if macro placement is both needed and feasible after it. We'd do this reporting for the user at the end of this function. With that said this should be resolved.